# Exercise 1 - Part 1: Word Embeddings

This notebook loads pre-trained word embedding vectors for `king`, `queen`, `dog`, `cat`, and `coffee`, then computes pairwise Euclidean distances.

Model used: `glove-wiki-gigaword-50` from `gensim.downloader`.

Important: the first run downloads the model, about 66 MB. After that, gensim caches it locally.

## 1. Install and import dependencies

If `gensim` is not installed in your Python environment, uncomment and run the install line below.

In [1]:
# Uncomment this line if gensim/scipy are missing:
# %pip install gensim scipy

from itertools import combinations

import gensim.downloader as api
import numpy as np

## 2. Load the pre-trained vectors

The selected model has 50-dimensional vectors. A larger model, such as Google's 300-dimensional Word2Vec, is also valid, but the exact distances will be different.

In [2]:
model_name = "glove-wiki-gigaword-50"
words = ["king", "queen", "dog", "cat", "coffee"]

model = api.load(model_name)
vectors = {word: model[word] for word in words}

print(f"Loaded model: {model_name}")
print(f"Words: {words}")

Loaded model: glove-wiki-gigaword-50
Words: ['king', 'queen', 'dog', 'cat', 'coffee']


## 3. What is the dimensionality?

The dimensionality is the length of each embedding vector. In this model, every selected word has a vector of length 50.

In [3]:
for word in words:
    print(f"{word}: {len(vectors[word])}")

king: 50
queen: 50
dog: 50
cat: 50
coffee: 50


## 4. Compute pairwise distances

I use Euclidean distance: the straight-line distance between two vectors. Smaller distance means the model places those words closer together in embedding space.

In [4]:
distances = []

for first, second in combinations(words, 2):
    distance = float(np.linalg.norm(vectors[first] - vectors[second]))
    distances.append((first, second, distance))

distances = sorted(distances, key=lambda item: item[2])

for first, second, distance in distances:
    print(f"{first}-{second}: {distance:.6f}")

dog-cat: 1.884603
king-queen: 3.477756
queen-cat: 4.968586
dog-coffee: 5.369360
queen-dog: 5.370511
cat-coffee: 5.472542
king-cat: 5.478320
king-dog: 5.891150
king-coffee: 6.544825
queen-coffee: 6.653850


## 5. Closest and furthest pairs

Closest pairs:

1. `dog` and `cat`, distance `1.884603`
2. `king` and `queen`, distance `3.477756`

Furthest pairs:

1. `queen` and `coffee`, distance `6.653850`
2. `king` and `coffee`, distance `6.544825`

In [5]:
closest = distances[0]
furthest = distances[-1]

print(f"Closest pair: {closest[0]}-{closest[1]} = {closest[2]:.6f}")
print(f"Furthest pair: {furthest[0]}-{furthest[1]} = {furthest[2]:.6f}")

Closest pair: dog-cat = 1.884603
Furthest pair: queen-coffee = 6.653850


## Explanation and reasoning

Word embeddings place words that appear in similar contexts near each other. The result makes intuitive sense:

- `dog` and `cat` are closest because both are common animals and household pets, so they appear in many similar contexts.
- `king` and `queen` are also close because both are royal titles and share many monarchy-related contexts.
- `coffee` is furthest from `queen` and `king` because it belongs to a drink/food context, while those words belong to a human/royalty context.

The important thinking point: the computer is not using dictionary definitions. It compares the learned numeric vectors, and those vectors summarize usage patterns from a large text corpus.